common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -22,64 +22,379 @@
+                Logger.errorPII(methodTag,"ApduException", e);
+            }
+        }
+    }
+
+    //Upon a positive button click in the cert picker, sets up the PIN prompt dialog.
+    public SmartcardCertPickerDialog.PositiveButtonListener getSmartcardCertPickerDialogPositiveButtonListener(final ClientCertRequest request) {
+        //final String methodTag = TAG + ":getSmartcardCertPickerDialogPositiveButtonListener";
+        return new SmartcardCertPickerDialog.PositiveButtonListener() {
+            @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+            @Override
+            public void onClick(final YubiKitCertDetails certDetails) {
+                //Need to prompt user for pin and verify pin. The positive button listener will handle the rest of the CBA flow.
+                final SmartcardPinDialog pinDialog = new SmartcardPinDialog(mActivity);
+                pinDialog.setPositiveButtonListener(getSmartcardPinDialogPositiveButtonListener(certDetails, pinDialog, request));
+                pinDialog.setNegativeButtonListener(new SmartcardPinDialog.NegativeButtonListener() {
+                    @Override
+                    public void onClick() {
+                        request.cancel();
+                        pinDialog.dismiss();
+                        mCurrentDialog = null;
+                    }
+                });
+                pinDialog.setCancelCbaCallback(new SmartcardPinDialog.CancelCbaCallback() {
+                    @Override
+                    public void onCancel() {
+                        request.cancel();
+                    }
+                });
+                pinDialog.show();
+                //Set currentDialog to pinDialog
+                synchronized (smartcardDialogLock) {
+                    mCurrentDialog = pinDialog;
+                }
+            }
+        };
+    }
+
+    //Upon a positive button click in the PIN prompt, verify the provided PIN and handle the results.
+    public SmartcardPinDialog.PositiveButtonListener getSmartcardPinDialogPositiveButtonListener(final YubiKitCertDetails certDetails, final SmartcardPinDialog pinDialog,final ClientCertRequest request) {
+        final String methodTag = TAG + ":getSmartcardPinDialogPositiveButtonListener";
+
+        return new SmartcardPinDialog.PositiveButtonListener() {
+            @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+            @Override
+            public void onClick(final String pin) {
+                synchronized (deviceLock) {
+                    // Need to request a SmartCardConnection in order to access certs on YubiKey.
+                    mDevice.requestConnection(UsbSmartCardConnection.class, new Callback<Result<UsbSmartCardConnection, IOException>>() {
+                        @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+                        @Override
+                        public void invoke(Result<UsbSmartCardConnection, IOException> value) {
```

Comment:
same here. handle ioexception if returned.

Reply:
^replied to other comment referring to this same issue.